corehq/motech/fhir/models.py

```diff
@@ -0,0 +1,169 @@
+                                      on_delete=models.CASCADE,
+                                      related_name='properties')
+
+    # `case_property`, `jsonpath` and `value_map` are set using the
+    # Data Dictionary UI.
+    case_property = models.ForeignKey(CaseProperty, on_delete=models.SET_NULL,
+                                      null=True, blank=True, default=None)
+    # Path to the FHIR resource property that corresponds with `case_property`
+    jsonpath = models.TextField(null=True, blank=True, default=None)
+    # Optional[dict] {CommCare value: FHIR value}
+    value_map = JSONField(null=True, blank=True, default=None)
+
+    # `value_source_config` is used when the Data Dictionary UI cannot
+    # do what you need.
+    value_source_config = JSONField(null=True, blank=True, default=None)
+
+    def save(self, *args, **kwargs):
+        if (
+            self.case_property
+            and self.case_property.case_type != self.resource_type.case_type
+        ):
+            raise ConfigurationError(
+                "Invalid FHIRResourceProperty: case_property case type "
+                f"'{self.case_property.case_type}' does not match "
+                f"resource_type case type '{self.resource_type.case_type}'.")
+        if (
+            (self.case_property or self.jsonpath or self.value_map)
+            and self.value_source_config
+        ):
+            raise ConfigurationError(
+                "Invalid FHIRResourceProperty: Unable to set "
+                "'value_source_config' when 'case_property', 'jsonpath' or "
+                "'value_map' are set.")
+        super().save(*args, **kwargs)
+
+    @property
+    def case_type(self) -> CaseType:
+        return self.resource_type.case_type
+
+    def get_value_source(self) -> ValueSource:
+        """
+        Returns a ValueSource for building FHIR resources.
+
+        e.g. You could build a FHIR resource as follows::
+
+            def build_fhir_resource(case, version=FHIR_VERSION_4_0_1):
+                case_type = CaseType.objects.get(
+                    domain=case.domain,
+                    name=case.type,
+                )
+                resource_type = FHIRResourceType.objects.get(
+                    case_type=case_type,
+                    fhir_version=version,
+                )
+
+                # CaseTriggerInfo packages data for use by ValueSource
+                prop_names = [p.name for p in case_type.properties]
+                info = CaseTriggerInfo(
+                    domain=domain,
+                    case_id=case.case_id,
+                    case_type=case.type,
+                    updates={p: case.get_case_property(p) for p in prop_names},
+                )
+
+                fhir_resource = resource_type.template or {}
+                for property in resource_type.properties:
+                    value_source = property.get_value_source()
+                    value_source.set_external_value(fhir_resource, info)
```

Comment:
is `set_external_value` defined somewhere? I couldn't find it in our codebase (which is generally why I'm not a huge fan of these types of comments, as they can become stale very quickly).

Reply:
Agreed. This will become a function once the jsonpath library supports building out `fhir_resource` instead of just updating existing documents, and then I will drop this comment.

... I'm also open to the idea of moving this from a comment to a function *before* the jsonpath library does what we need. At least then there is no chance it could get stale.

Reply:
@proteusvacuum @mkangia I opened a draft PR for the code that implements `set_external_value()` (and drops this docstring): <https://github.com/dimagi/commcare-hq/pull/29111>